src/main/ngapp/accio/src/app/editor/magic-wand.service.ts

```diff
@@ -0,0 +1,169 @@
+    if (!isValid) {
+      // Automatically not in mask b/c failed vailidity test
+      return false;
+    }
+
+    visited.add(this.coordToDataArrayIndices(curX, curY, imgWidth)[0]);
+
+    // Get array of attributes of current pixel
+    let curPixel: Array<number> = this.dataArrayToRGBA(imgData, curX, curY);
+
+    // All attributes of the pixel (R,G,B, and A) must be within tolerance level
+    for (let i = 0; i < 4; i++) {
+      let upperTolerance: boolean = curPixel[i] > originalPixel[i] + tolerance;
+      let lowerTolerance: boolean = curPixel[i] < originalPixel[i] - tolerance;
+      if (upperTolerance || lowerTolerance) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  // Check if @pixelCoord is in bounds and makes sure it's not a repeat coord
+  getIsValid(imgWidth: number, imgHeight: number, curX: number, curY: number,
+    visited: Set<number>): boolean {
+    if (this.isInBounds(imgWidth, imgHeight, curX, curY) &&
+      this.notVisited(imgWidth, curX, curY, visited)) {
+      return true;
+    }
+
+    return false;
+  }
+
+  // Check bounds of indexing for img dimensions
+  isInBounds(imgWidth: number, imgHeight: number, curX: number, curY: number): boolean {
+    let yOutOfBounds: boolean = curY < 0 || curY > imgHeight - 1;
+    let xOutOfBounds: boolean = curX < 0 || curX > imgWidth - 1;
+    if (yOutOfBounds || xOutOfBounds) {
+      return false;
+    }
+
+    return true;
+  }
+
+  // Checks if pixel has been visited already
+  notVisited(imgWidth:number, curX: number, curY: number,
+    visited: Set<number>): boolean {
+    // Do not push repeat coords to heap
+    let index: number =
+      this.coordToDataArrayIndices(curX, curY, imgWidth)[0];
+    if (visited.has(index)) {
+      return false;
+    }
+
+    return true;
+  }
+
+
+  // Return pixel attributes of @imgData at [@xCoord, @yCoord] as [R, G, B, A]
+  dataArrayToRGBA(imgData: ImageData, xCoord: number, yCoord: number):
+    Array<number> {
+    // Unpack imgData for readability
+    let data: Uint8ClampedArray = imgData.data;
+    let imgWidth: number = imgData.width;
+
+    // Pixel attributes in imgData are organized adjacently in a 1-D array
+    let pixelIndex: Array<number> =
+      this.coordToDataArrayIndices(xCoord, yCoord, imgWidth);
+    let red: number = data[pixelIndex[0]];
+    let green: number = data[pixelIndex[1]];
+    let blue: number = data[pixelIndex[2]];
+    let alpha: number = data[pixelIndex[3]];
+    // Store original pixel's attributes
+    return [red, green, blue, alpha];
+  }
+
+  // Convert coord [@x, @y] (2-D) to indexing style of DataArray (1-D)
+  /**Returns [indexOfR, indexOfG, indexOfB, indexOfA]
+   * Important: Returns array of number representing indices.
+   * First element contains value of index for attribute: red and so on
+   * */
+  coordToDataArrayIndices(x: number, y: number, width: number): Array<number> {
```

Comment:
In every caller except one, you are just using the first element of the returned array. I would suggest having this function just return the first index, and then doing the +1, +2, etc in that one call site.